test/distributed/fsdp/test_fsdp_optim_state.py

```diff
@@ -0,0 +1,435 @@
+# Owner(s): ["oncall: distributed"]
+
+import sys
+from typing import Any, Dict, List, Type
+
+import torch
+from torch import distributed as dist
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.fully_sharded_data_parallel import (
+    OPTIM_TARGET_RANK,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest
+from torch.testing._internal.common_utils import (
+    TEST_WITH_DEV_DBG_ASAN,
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class Bias(torch.nn.Module):
```

Comment:
These `nn.Module` classes (`Bias`, `AffineA`, `AffineB`, `AffineModel`) can be moved to `common_fsdp.py` and used for as an example model with non-`FSDP` root and nontrivial nesting if you guys want.

`AffineModel` was written to be compliant with the `common_fsdp.py` models; i.e. it has `get_input()`, `get_loss()`, and `run_backward()`.